File: src/net/dinkla/raytracer/objects/acceleration/kdtree/Statistics.java
package net.dinkla.raytracer.objects.acceleration.kdtree;

import net.dinkla.raytracer.math.Histogram;

/**
* Created by IntelliJ IDEA.
* User: jorndinkla
* Date: 11.06.2010
* Time: 11:43:35
* To change this template use File | Settings | File Templates.
*/
public class Statistics {
    int numObjects;
    int numInner;
    int numLeafs;
    int numObjectsInLeafs;

    Histogram depthsLeafs;
    Histogram depthsNodes;
    Histogram numChildrenInLeafs;
    Histogram numObjectsShared;

    public Statistics() {
        numInner = 0;
        numLeafs = 0;
        numObjectsInLeafs = 0;

        depthsNodes = new Histogram();
        depthsLeafs = new Histogram();
        numChildrenInLeafs = new Histogram();
    }

    public static void cs(AbstractNode node, Statistics s, int depth) {
        if (node instanceof Leaf) {
            Leaf n = (Leaf) node;
            s.numLeafs++;
            s.depthsLeafs.add(depth);
            int sz = n.size();
            s.numChildrenInLeafs.add(sz);
            s.numObjectsInLeafs += sz;
        } else if (node instanceof InnerNode) {
            InnerNode n = (InnerNode) node;
            s.numInner++;
            s.depthsNodes.add(depth);
            cs(n.left, s, depth+1);
            cs(n.right, s, depth+1);
        }
    }

    public static void statistics(KDTree tree) {

        Statistics s = new Statistics();

        cs(tree.root, s, 0);

        System.out.println("num objects=" + tree.size());
        System.out.println("inner nodes=" + s.numInner);
        System.out.println("leaf nodes=" + s.numLeafs);
        System.out.println("objects in leafs=" + s.numObjectsInLeafs);

        System.out.println("depthsLeafs");
        s.depthsLeafs.println();

        System.out.println("depthsNodes");
        s.depthsNodes.println();


        System.out.println("numChildrenInLeafs");
        s.numChildrenInLeafs.println();

        // Anzahl Innernodes + Anzahl Leafs
        // Anzahl Objekte

        // maxDepth
        // minDepth Leaf
        // Histogram Knoten in Leafs
        // Histogram Object in Knoten (shared objects)
        // Histogramm Knoten pro Tiefe

    }
}

File: src/main/kotlin/net/dinkla/raytracer/objects/acceleration/kdtree/Statistics.kt
package net.dinkla.raytracer.objects.acceleration.kdtree

import net.dinkla.raytracer.math.Histogram

class Statistics {
    internal var numObjects: Int = 0
    internal var numInner: Int = 0
    internal var numLeafs: Int = 0
    internal var numObjectsInLeafs: Int = 0

    internal var depthsLeafs: Histogram
    internal var depthsNodes: Histogram
    internal var numChildrenInLeafs: Histogram
    internal var numObjectsShared: Histogram? = null

    init {
        numInner = 0
        numLeafs = 0
        numObjectsInLeafs = 0

        depthsNodes = Histogram()
        depthsLeafs = Histogram()
        numChildrenInLeafs = Histogram()
    }

    companion object {

        fun cs(node: AbstractNode?, s: Statistics, depth: Int) {
            if (node is Leaf) {
                val n = node as Leaf?
                s.numLeafs++
                s.depthsLeafs.add(depth)
                val sz = n!!.size()
                s.numChildrenInLeafs.add(sz)
                s.numObjectsInLeafs += sz
            } else if (node is InnerNode) {
                val n = node as InnerNode?
                s.numInner++
                s.depthsNodes.add(depth)
                cs(n!!.left, s, depth + 1)
                cs(n.right, s, depth + 1)
            }
        }

        fun statistics(tree: KDTree) {
            val s = Statistics()
            cs(tree.root, s, 0)

            println("num objects=" + tree.size())
            println("inner nodes=" + s.numInner)
            println("leaf nodes=" + s.numLeafs)
            println("objects in leafs=" + s.numObjectsInLeafs)

            println("depthsLeafs")
            s.depthsLeafs.println()

            println("depthsNodes")
            s.depthsNodes.println()

            println("numChildrenInLeafs")
            s.numChildrenInLeafs.println()

            // Anzahl Innernodes + Anzahl Leafs
            // Anzahl Objekte

            // maxDepth
            // minDepth Leaf
            // Histogram Knoten in Leafs
            // Histogram Object in Knoten (shared objects)
            // Histogramm Knoten pro Tiefe
        }
    }
}
